### PR TITLE
Added des to  rlib & the options to be used by rahash2 (or anything else)

### PR DESCRIPTION
**Pull request by wargio**
DES for everyone!!! :D

regression tests: https://github.com/radare/radare2-regressions/pull/684

### CODE REVIEW

**Comment by Maijin** [comment]
Does it works with strings ? If yes please add test for that too 👍

**Comment by wargio** [comment]
Added.

File: libr/crypto/p/crypto_des.c [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+    *keyhi = perm >> 4;
+}
+
+// first permutation of the block
+static void des_permute_block0  (ut32 *blocklo, ut32 *blockhi) {
+    ut32 lo, hi, perm;
+    if (!blocklo || !blockhi) return;
+    lo = *blocklo;
+    hi = *blockhi;
+	perm = ((lo >> 4) ^ hi) & 0x0F0F0F0F;
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: use tabs

File: libr/crypto/p/crypto_des.c [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    *keylo |= ((*keyhi << 4) & 0x000FF000);
+    *keylo |= ((*keyhi >> 12) & 0x00000FF0);
+    *keylo |= ((*keyhi >> 28) & 0x0000000F);
+    *keyhi = perm >> 4;
+}
+
+// first permutation of the block
+static void des_permute_block0  (ut32 *blocklo, ut32 *blockhi) {
+    ut32 lo, hi, perm;
+    if (!blocklo || !blockhi) return;
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: add braces

File: libr/crypto/p/crypto_des.c [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    //last permutation
+    des_permute_block1(&st->bufhi, &st->buflo);
+ 
+    //result
+    wbe32(output + 0, st->bufhi);
+    wbe32(output + 4, st->buflo);
+
+    return true;
+}
+
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: empty line

File: libr/crypto/p/crypto_des.c [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+static struct des_state st;
+static int flag = 0;
+
+static int des_set_key (RCrypto *cry, const ut8 *key, int keylen, int mode, int direction) {
+    ut32 keylo, keyhi, i;
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: use only tabs to indent

File: libr/crypto/p/crypto_des.c [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+	.name = "des-ecb",
+	.set_key = des_set_key,
+	.get_key_size = des_get_key_size,
+	.use = des_use,
+	.update = update,
+	.final = final
+};
+
+#ifndef CORELIB
+struct r_lib_struct_t radare_plugin = {
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: s/struct r_lib_struct_t/RLibStruct/

File: libr/crypto/p/crypto_des.c [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+    wbe32(output + 4, st->buflo);
+
+    return true;
+}
+
+
+static struct des_state st;
+static int flag = 0;
+
+static int des_set_key (RCrypto *cry, const ut8 *key, int keylen, int mode, int direction) {
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: s/int/bool/

File: libr/crypto/p/crypto_des.c [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+
+static int des_get_key_size (RCrypto *cry) {
+	return DES_KEY_SIZE;
+}
+
+static bool des_use (const char *algo) {
+	return !strcmp (algo, "des-ecb");
+}
+
+static int update (RCrypto *cry, const ut8 *buf, int len) {
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: s/int/bool/

File: libr/crypto/p/crypto_des.c [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+	// Pad to the block size, do not append dummy block
+	const int diff = (BLOCK_SIZE - (len % BLOCK_SIZE)) % BLOCK_SIZE;
+	const int size = len + diff;
+	const int blocks = size / BLOCK_SIZE;
+
+	ut8 *const obuf = calloc (1, size);
+	if (!obuf) return false;
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: add braces

File: libr/crypto/p/crypto_des.c [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+
+	ut8 *const ibuf = calloc (1, size);
+	if (!ibuf) {
+		free (obuf);
+		return false;
+	}
+
+	memset (ibuf, 0, size);
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: memset (ibuf + len, 0, (size-len))

File: libr/crypto/p/crypto_des.c [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+static bool des_use (const char *algo) {
+	return !strcmp (algo, "des-ecb");
+}
+
+static int update (RCrypto *cry, const ut8 *buf, int len) {
+	// Pad to the block size, do not append dummy block
+	const int diff = (BLOCK_SIZE - (len % BLOCK_SIZE)) % BLOCK_SIZE;
+	const int size = len + diff;
+	const int blocks = size / BLOCK_SIZE;
+
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: ensure len is > 0

File: libr/crypto/p/crypto_des.c [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+	memset (ibuf, 0, size);
+	memcpy (ibuf, buf, len);
+// got it from AES, should be changed??
+// Padding should start like 100000...
+//	if (diff) {
+//		ibuf[len] = 8; //0b1000;
+//	}
+
+	int i;
+	if (flag == 0) {
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: rename `flag` to `bool doEncrypt`, or something like that.

File: libr/crypto/p/crypto_des.c [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+	keyhi = be32(key + 4);
+
+	st.key_size = DES_KEY_SIZE;
+	st.rounds = 16;
+	flag = direction;
+    // key permutation to derive round keys
+	des_permute_key (&keylo, &keyhi);
+    for (i = 0; i < 16; ++i) {
+        // filling round keys space
+        des_round_key(i, &st.keylo[i], &st.keyhi[i], &keylo, &keyhi);
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: missing space before (

File: libr/crypto/p/crypto_des.c [hunk 13]
```diff
@@ -0,0 +1,392 @@
...
+static struct des_state st;
+static int flag = 0;
+
+static int des_set_key (RCrypto *cry, const ut8 *key, int keylen, int mode, int direction) {
+    ut32 keylo, keyhi, i;
+	if (keylen != DES_KEY_SIZE) {
+		return false;
+	}
+    // splitting the key in hi & lo
+	keylo = be32(key);
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: missing space before (

File: libr/crypto/p/crypto_des.c [hunk 14]
```diff
@@ -0,0 +1,392 @@
...
+ 
+	//result
+    wbe32(output + 0, st->bufhi);
+    wbe32(output + 4, st->buflo);
+
+    return true;
+}
+
+static int des_decrypt (struct des_state *st, const ut8 *input, ut8 *output) {
+    if (!st || !input || !output) return false;
```
Comment (radare) [inline libr/crypto/p/crypto_des.c]: add braces here and one space before every ( except in function signature definition

**Comment by wargio** [comment]
Ok, I'll add all those..  I thought spaces were removed since I did a replace all with \t...

File: libr/crypto/p/crypto_des.c (hunk 9 above)
Comment (wargio) [inline libr/crypto/p/crypto_des.c]: Why?

**Comment by wargio** [comment]
travis failed on this last commit due regression test `19` on `rahash2`, since it outputs `des` in the `rehash2 -L`

**Comment by Maijin** [comment]
yes that's fine

**Comment by alvarofe** [comment]
Tests are broken please review, everything has been merged and master is red because of it.

**Pull request merged**